pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -188,97 +231,110 @@ func (cmvt *cdcMixedVersionTester) waitForResolvedTimestamps() versionStep {
-
-		// we could just return the error here and let `Wait` return the
-		// error. However, calling t.Fatal directly lets us stop the test
-		// earlier
-		cmvt.monitor.Go(func(ctx context.Context) error {
-			defer close(cmvt.validatorFinished)
-			consumer, err := cmvt.kafka.consumer(ctx, targetTable)
-			if err != nil {
-				t.Fatal(err)
+// runKafkaConsumer continuously reads from the kafka consumer and places
+// messages in a buffer. It expects to be run in a goroutine and does not
+// rely on any DB connections.
+func (cmvt *cdcMixedVersionTester) runKafkaConsumer() mixedversion.UserFunc {
+	return func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		everyN := log.Every(30 * time.Second)
+
+		// This runs until the test finishes, which will be signaled via
+		// context cancellation. We rely on consumer.Next() to check
+		// the context.
+		for {
+			m := cmvt.kafka.consumer.Next(ctx)
+			if m == nil {
+				// this is expected to happen once the test has finished and
+				// Kafka is being shut down. If it happens in the middle of
+				// the test, it will eventually time out, and this message
+				// should allow us to see that the validator finished
+				// earlier than it should have
+				l.Printf("end of changefeed")
+				return nil
 			}
-			defer consumer.Close()
 
-			db := u.conn(ctx, t, node)
-			if _, err := db.Exec(
-				"CREATE TABLE fprint (id INT PRIMARY KEY, balance INT, payload STRING)",
-			); err != nil {
-				t.Fatal(err)
+			// Forward resolved timetsamps to "heartbeat" that the changefeed is running.
+			if len(m.Key) <= 0 {
+				_, resolved, err := cdctest.ParseJSONValueTimestamps(m.Value)
+				if err != nil {
+					return err
+				}
+				cmvt.timestampResolved()
+
+				if everyN.ShouldLog() {
+					l.Printf("latest resolved timestamp %d behind realtime", timeutil.Since(resolved.GoTime()))
+				}
 			}
 
-			getConn := func(node int) *gosql.DB { return u.conn(ctx, t, node) }
-			fprintV, err := cdctest.NewFingerprintValidator(db, tableName, `fprint`, consumer.partitions, 0)
-			if err != nil {
-				t.Fatal(err)
+			if len(cmvt.kafka.buf) == kafkaBufferMessageSize {
+				return errors.Newf("kafka message buffer limit %d exceeded", kafkaBufferMessageSize)
 			}
-			fprintV.DBFunc(cmvt.cdcDBConn(getConn))
-			validators := cdctest.Validators{
-				cdctest.NewOrderValidator(tableName),
-				fprintV,
+			cmvt.kafka.buf = append(cmvt.kafka.buf, m)
+		}
+	}
+}
+
+// validate reads rows from the kafka buffer and submits them to the validator
+// to be validated.
+func (cmvt *cdcMixedVersionTester) validate() mixedversion.UserFunc {
+	return func(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper) error {
+		// NB: The caller is responsible for making sure this function is run
+		// when nodes are running.
+		validator := cmvt.validator.UnsafeGetValidator()
+		for _, m := range cmvt.kafka.buf {
```

Comment:
Don't you need to protect this from concurrent writes happening in `runKafkaConsumer`?